src/Collections-Sequenceable/Array.class.st

```diff
@@ -136,13 +136,48 @@ Array class >> new: sizeRequested [
 	^ self basicNew: sizeRequested  "Exceptional conditions will be handled in basicNew:"
 ]
 
+{ #category : #converting }
+Array >> as: aSimilarClass [
+
+	"Specific classes can redefine newFromArray: by a faster implementation if they want.
+
+	Note: see Collection class>>newFromArray: for the rationale."
+
+	aSimilarClass == self class ifTrue: [ ^ self ].
+	^ aSimilarClass newFromArray: self
+]
+
 { #category : #converting }
 Array >> asArray [
 	"Answer with the receiver itself."
 
 	^ self
 ]
 
+{ #category : #converting }
+Array >> asDictionary [
+
```

Comment:
I would not add asDictionary because it is really strange to me and can only work if we have binding in the array.

Reply:
I'm not against this change per se.
@Ducasse , asDictionary already exists. And works only for collections with associations.
So, this change is semantically equivalent.

Of course, the rationale behind this is speed, but if the change is questioned for Dictionaries, then all the rest of the changes are too. The question is: Do we want `{ a->b. c->d } asDictionary` to be preferred to  `Dictionary new at: a put: b; at: c put: d; yourself`?.

Regarding performance, however, this only cuts one message send, right? Instead of

```
anArray asDictionary
anArray as: Dictionary
Dictionary newFromArray: anArray
...
```

It would be:
```
anArray asDictionary
Dictionary newFromArray: anArray
...
```

right?
Is there much to gain in here?

Reply:
Since there was an existing `asDictionary` in Array, I just adapted it to use the fast path from `Collection class>>newFromArray:` and be equivalent with `as: Dictionnary` (as `Array>>as:` uses the fast path by default).

Note: see the doc of `Collection class>>newFromArray:` for where is the gain

Reply:
I forgot the rule of  ``do not comment before your first coffee''.

It seems that there was in fact no previous definition of `asDictionary` therefore I'm not sure why I added it.
I will remove it since there is almost no gain as noted @guillep

Reply:
There is #asDictionary in Pharo (on Collection). It has 90 senders.

Reply:
I propose to make *this* PR solely about speeding up what we have and do change as part of another step.

Reply:
I removed asDictionary